packages/dds/merge-tree/src/segmentPropertiesManager.ts

```diff
@@ -45,133 +47,318 @@ export function copyPropertiesAndManager(
+				iterateListValues(pending.remote.first, (n) => {
+					if (n.data.seq <= msn) {
+						n.list?.remove(n);
+						return true;
+					}
+					return false;
+				}),
+			);
+			if (pending.local.empty && pending.remote.empty) {
+				this.changes.delete(key);
+			}
+		}
 	}
 
+	/**
+	 * Copies properties to another manager.
+	 * This method copies the properties and their changes from the current manager to the destination manager.
+	 *
+	 * @param oldProps - The old properties to be copied.
+	 * @param dest - The destination object containing properties and property manager.
+	 */
 	public copyTo(
-		oldProps: PropertySet,
-		newProps: PropertySet | undefined,
-		newManager: PropertiesManager,
-	): PropertySet | undefined {
-		if (oldProps) {
-			// eslint-disable-next-line no-param-reassign
-			newProps ??= createMap<unknown>();
-			if (!newManager) {
-				throw new Error("Must provide new PropertyManager");
-			}
-			extend(newProps, oldProps);
+		oldProps: PropertySet | undefined,
+		dest: {
+			properties?: PropertySet;
+			propertyManager?: PropertiesManager;
+		},
+	): void {
+		const newManager = (dest.propertyManager ??= new PropertiesManager());
+		dest.properties = clone(oldProps);
+		for (const [key, { local, remote, msnConsensus }] of this.changes.entries()) {
+			newManager.changes.set(key, {
+				msnConsensus,
+				remote: new DoublyLinkedList(remote.empty ? undefined : remote.map((c) => c.data)),
+				local: new DoublyLinkedList(local.empty ? undefined : local.map((c) => c.data)),
+			});
+		}
+	}
 
-			if (this.pendingKeyUpdateCount) {
-				newManager.pendingKeyUpdateCount = clone(this.pendingKeyUpdateCount);
+	/**
+	 * Gets properties at a specific sequence number.
+	 * This method retrieves the properties at the given sequence number.
+	 * This is only needed to support emitting snapshots in the legacy format.
+	 * If we remove the ability to emit the legacy format, we can remove this method, along with the need to track remote changes at all.
+	 *
+	 * @param oldProps - The old properties to be retrieved.
+	 * @param sequenceNumber - The sequence number to get properties at.
+	 * @returns The properties at the given sequence number.
+	 */
+	public getAtSeq(
+		oldProps: MapLike<unknown> | undefined,
+		sequenceNumber: number,
+	): MapLike<unknown> {
+		const properties: MapLike<unknown> = { ...oldProps };
+		for (const [key, changes] of this.changes) {
+			properties[key] = computePropertyValue(
+				changes.msnConsensus,
+				iterateListValues(changes.remote.first, (c) => c.data.seq <= sequenceNumber),
+			);
+			if (properties[key] === null) {
+				// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+				delete properties[key];
 			}
 		}
-		return newProps;
+		return properties;
 	}
 
 	/**
-	 * Determines if all of the defined properties in a given property set are pending.
+	 * Checks if there are pending properties.
+	 * This method checks if there are any pending properties that need to be acknowledged.
```

Comment:
I liked the previous doc better. Whatever you go with, it should emphasize that it returns true iff *all* property keys in the `props` parameter are pending. And this looks like the semantics that the only call site wants (it's only used in an assert in `client.ts` when acking an annotate op)